x-pack/test/security_api_integration/tests/features/deprecated_features.ts

```diff
@@ -0,0 +1,188 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { FeatureKibanaPrivilegesReference, KibanaFeatureConfig } from '@kbn/features-plugin/common';
+import { expect } from 'expect';
+import { FtrProviderContext } from '../../ftr_provider_context';
+
+function collectReplacedByForFeaturePrivileges(
+  feature: KibanaFeatureConfig
+): Array<[string, readonly FeatureKibanaPrivilegesReference[]]> {
+  const privilegesToReplace = [] as Array<[string, readonly FeatureKibanaPrivilegesReference[]]>;
+  if (feature.privileges) {
+    const allReplacedBy = feature.privileges.all.replacedBy ?? [];
+    const readReplacedBy = feature.privileges.read.replacedBy ?? [];
+    privilegesToReplace.push([
+      'all',
+      'default' in allReplacedBy ? allReplacedBy.default : allReplacedBy,
+    ]);
+    privilegesToReplace.push([
+      'minimal_all',
+      'minimal' in allReplacedBy ? allReplacedBy.minimal : allReplacedBy,
+    ]);
+    privilegesToReplace.push([
+      'read',
+      'default' in readReplacedBy ? readReplacedBy.default : readReplacedBy,
+    ]);
+    privilegesToReplace.push([
+      'minimal_read',
+      'minimal' in readReplacedBy ? readReplacedBy.minimal : readReplacedBy,
+    ]);
+  }
+
+  for (const subFeature of feature.subFeatures ?? []) {
+    for (const group of subFeature.privilegeGroups) {
+      for (const subFeaturePrivilege of group.privileges) {
+        privilegesToReplace.push([subFeaturePrivilege.id, subFeaturePrivilege.replacedBy ?? []]);
+      }
+    }
+  }
+
+  return privilegesToReplace;
+}
+
+function getActionsToReplace(actions: string[]) {
+  // The `ui:`-prefixed actions are special since they are prefixed with a feature ID, and do not need to be replaced
+  // like any other privileges.
+  // TODO: What about `alerting`?
```

Comment:
> // TODO: What about `alerting`?

What about it? 😄

Reply:
Curious about this as well!

Reply:
Still need to validate 🙂 I'll tag you once I add tests for that

Reply:
Okay, I checked how migration would work for Cases and Alerting custom RBAC. For Cases, nothing special is needed since the actions we generate for the cases part of the feature and feature privilege definitions aren’t tied to a feature ID. You can see the tests I’ve added here: https://github.com/elastic/kibana/pull/186800/files#diff-5dca65d5b076d6486e7fe8b00c6903d22475f8b3ff3a1ac9fd275d8f0126a0baR246-R342

Alerting RBAC is trickier since it’s hard-wired to a feature ID, and the actions generated for the same rule or alert type, granted by different features, are different and hence rules and alerts can "belong" only to a specific feature. This means that to allow users with roles referencing privileges from deprecated features that include alerting, we need to complement the alerting-related actions of the deprecated feature’s privileges with those from the features that replace them — just like we do for UI actions, which are also hard-wired to feature IDs. I’ve added this change [here](https://github.com/elastic/kibana/pull/186800/files#diff-edc7256c39f51cfdd636c8bad9e76727bb8d56c3eda637df4968f24a4880cd3aR106) and covered it with tests [here](https://github.com/elastic/kibana/pull/186800/files#diff-e32b3f76d0706db202b5b3069c3649f135ef2ba594615153cc31509637d6a873R610-R612) and [here](https://github.com/elastic/kibana/pull/186800/files#diff-5dca65d5b076d6486e7fe8b00c6903d22475f8b3ff3a1ac9fd275d8f0126a0baR344-R451).

I couldn’t add tests for the reverse scenario, where users with the new/replacement privileges try to access rules and alerts created via deprecated features, since it requires SO migration for the related rules and alerts to update the “consumer” field to point to a new feature. This will change in [this PR](https://github.com/elastic/kibana/pull/183756) (targeting 8.17.0), where feature IDs will be decoupled from the “consumers” field in rules and alerts. I’m going to work with @cnasikas to ensure this change is compatible with the privilege deprecation framework. No one is planning to deprecate features with alerting in 8.16, so handling this in 8.17 sounds reasonable to me. I’ve commented out the tests that we need to uncomment in 8.17 [here](https://github.com/elastic/kibana/pull/186800/files#diff-5dca65d5b076d6486e7fe8b00c6903d22475f8b3ff3a1ac9fd275d8f0126a0baR422-R450).

Reply:
Extended alerting tests in https://github.com/elastic/kibana/pull/183756/commits/a200984f52c3353c41f23b61c5f50ecde6f650f3